### PR TITLE
Code Execution

### LINKED ISSUE #54: N3: Code Execution

### PR DESCRIPTION
**Pull request by Nafeij**
- New models under `mongo` schema: `Environment` and `TestCase` + relevant routes under `QuestionRouter`
- Independent Judge0 server w/ endpoint + `JudgeRouter`
- New frontend for testing: `Test Code`

Currently a frontend for adding Environments and Cases to Questions is not planned. Suggest manipulating the DB directly through etc DataGrip. Otherwise, some new example data is provided in `data.zip`.

Closes #54

### CODE REVIEW

File: src/server/api/routers/judge.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  compiler_options: z.string().optional(),
+  stdin: z.string().optional(),
+  expected_output: z.string().optional(),
+  cpu_time_limit: z.number().optional(),
+  cpu_extra_time: z.number().optional(),
+  memory_limit: z.number().optional(),
+});
+
+export const judgeRouter = createTRPCRouter({
+  getLanguages: maintainerProcedure.query(async () => {
```
Comment (Gabau) [inline src/server/api/routers/judge.ts:44]: Why is this a maintainer procedure? Shouldn't it be protected so any user can get the languages? Same for the others

File: src/server/api/routers/judge.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    return axios
+      .post(
+        "http://localhost:2358/submissions/?base64_encoded=false&wait=true",
+        input,
+      )
+      .then((res) => {
+        return res.data as Output;
+      });
+  }),
+  runTestCase: protectedProcedure.input(testCaseExecutionObject).mutation(async ({ ctx, input }) => {
```
Comment (Gabau) [inline src/server/api/routers/judge.ts]: Is there a way to pass in time limit and memory limits or it is set in the conf?

**Review by Gabau — APPROVED** [review]
Some minor comments. The code execution works great! Only tested on cpp and python tho

File: src/server/api/routers/judge.ts (hunk 1 above)
Comment (Nafeij) [reply src/server/api/routers/judge.ts:44]: The idea was to restrict the scope of actions such that normal users can only run code in the context of test cases. User clients should fetch languages only wrt those available for the particular question (i.e. via the `question` route).

File: src/server/api/routers/judge.ts (hunk 2 above)
Comment (Nafeij) [reply src/server/api/routers/judge.ts]: I've added the relevant fields in the TestCase modal

**Pull request merged**